openedx/openedx-learning
# feat: adds Collection.key

## Description

Adds a new required field to the Collection model: `key` (string), which uniquely identifies a Collection within a Learning Package.
Existing Collection keys are initialized with a random string.

## Supporting information

Part of : https://github.com/openedx/modular-learning/issues/226

Private-ref: [FAL-3783](https://tasks.opencraft.com/browse/FAL-3783)

## Testing instructions

See https://github.com/openedx/edx-platform/pull/35321 for full testing instructions.

Testing migrations:

In an openedx-learning virtual environment, run:

0. Create an openedx-learning virtual environment
1. Install openedx-learning @ `main`, and run `python3 manage.py migrate`
2. Create a collection `python3 manage.py shell`:
   ```
   import openedx_learning.api.authoring as api
   learning_package = api.create_learning_package("LP1", "Learning Package 1")
   for x in range(0, 11):
       api.create_collection(learning_package.id, title=f"Collection {x}", created_by=None)
       # <Collection> (1:Collection 1)
       # <Collection> (2:Collection 2)
       ...
       # <Collection> (10:Collection 10)
   ```
2. Install openedx-learning @ `open-craft/jill/collection-key` 
3. Run `python3 manage.py migrate`
4. Check that a key was generated for your collection  `python3 manage.py shell`:
   ```
   import openedx_learning.api.authoring as api
   learning_package = api.get_learning_package_by_key("LP1")
   collections = api.get_collections(learning_package.id)
   for collection in collections.all():
       print(collection.key)
       # gy19KxxV9xrjEMmR4TRNeE2p8sEbQfetVjZPzKPUbHURAgAgJY
       # RNeE2p8sEbQfetVjZPzKPUbgy19KxxV9xrjEMmR4TRNeE2p8sEb
       ...
   ```
